lib/kiev/param_filter.rb

```diff
@@ -17,14 +26,21 @@ def self.filter(params, filtered_params, ignored_params)
         end
 
         acc[key] =
-          if filtered_params.include?(key) && !value.is_a?(Hash)
+          if filtered_params.include?(key.to_s) && !value.is_a?(Hash)
             FILTERED
           elsif value.is_a?(Hash)
-            filter(value, filtered_params, ignored_params)
+            call(value)
           else
             value
           end
       end
     end
+
+    private
+    attr_reader :filtered_params, :ignored_params
+
+    def normalize(params)
+      Set.new(params.map(&:to_s))
```

Comment:
While this does improve search speed for bigger keysets, it also creates the same sets on every run, potentially negating the win for smaller keysets. Caching would be an easy answer, but I'd even go as far as to make the **configuration itself** contain sets of strings rather then them being derivatives of the original configuration.

Reply:
That was my initial idea. I did not follow it because otherwise the interface of adding custom filtering parameters would be changed significantly. Set doesn't have methods like `+=`, `append` that are commonly used on arrays. Moreover, overriding default params would be even more unobvious (`config.filtered_params = Set.new([:param])` instead of `config.filtered_params = [:param]`).

And in addition, while I was adjusting this code I was asking myself why `ParamFilter.filter(params, filtered, ignored)` method expects `filtered` and `ignored` for every call instead of fetching this configuration directly from the `Config` instance every time? Probably there is an intention of such behavior, to give an ability to use `Kiev::ParamFilter` directly.

Because of all mentioned reasons, I'd keep caching in `Kiev::ParamFilter` directly.

Reply:
Yeah, this'll change the interface. The new interface ain't bad though:
```
s = Set[foo]        # {foo}
s.add(bar)          # {foo, bar}
s.replace([baz])    # {baz}
s << foo            # {foo baz}
s += [bar]          # {foo bar baz}
s.merge([bar, qux]) # {foo bar baz qux}
```

Reply:
But on second thought, maybe let's focus this PR on compliance and address a performance problem separately.